backend/CircleForms/Startup.cs

```diff
@@ -1,182 +1,201 @@
+        services.Configure<OsuApiConfig>(config);
+        services.Configure<SuperAdminsId>(Configuration.GetSection("SuperAdmins"));
+        services.Configure<StaticFilesConfig>(Configuration.GetSection("StaticFiles"));
+
+        services.AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblyContaining<Startup>());
+
+        TypeAdapterConfig.GlobalSettings.Scan(Assembly.GetEntryAssembly()!);
+        TypeAdapterConfig.GlobalSettings.Compiler = x => x.CompileFast();
+        services.AddTransient<IMapper, Mapper>();
+
+        services.AddAuthentication("InternalCookies")
+            .AddCookie("InternalCookies", options =>
+            {
+                // set some paths to empty to make auth not redirect API calls
+                options.LoginPath = string.Empty;
+                options.AccessDeniedPath = string.Empty;
+                options.LogoutPath = string.Empty;
+                options.Cookie.Path = "/";
+                options.SlidingExpiration = true;
+                options.Events.OnValidatePrincipal = context =>
+                {
+                    var name = context.Principal?.Identity?.Name;
+                    if (string.IsNullOrEmpty(name) || !long.TryParse(name, out _))
+                    {
+                        context.RejectPrincipal();
+                        context.Response.StatusCode = StatusCodes.Status401Unauthorized;
+                    }
+
+                    return Task.CompletedTask;
+                };
+
+                static Task UnauthorizedRedirect(RedirectContext<CookieAuthenticationOptions> context)
+                {
+                    context.Response.StatusCode = StatusCodes.Status401Unauthorized;
+
+                    return Task.CompletedTask;
+                }
+
+                options.Events.OnRedirectToLogin = UnauthorizedRedirect;
+                options.Events.OnRedirectToAccessDenied = UnauthorizedRedirect;
+            })
+            .AddCookie("ExternalCookies")
+            .AddOAuth("osu", options =>
+            {
+                options.SignInScheme = "ExternalCookies";
+
+                options.TokenEndpoint = "https://osu.ppy.sh/oauth/token";
+                options.AuthorizationEndpoint = "https://osu.ppy.sh/oauth/authorize";
+                options.ClientId = config["ClientID"];
+                options.ClientSecret = config["ClientSecret"];
+                options.CallbackPath = config["CallbackUrl"];
+                options.Scope.Add("public");
+
+                options.SaveTokens = true;
+
+                options.Validate();
+            });
+
+        services.AddSingleton<IOsuApiProvider, OsuApiProvider>();
+        services.AddTransient<IUserRepository, UserRepository>();
+        services.AddTransient<IPostRepository, PostRepository>();
+        services.AddTransient<IAnswerService, AnswerService>();
+        services.AddTransient<IPublishService, PublishService>();
+        services.AddTransient<PostsService>();

```

Comment:
Might as well add an interface for `PostsService`

Reply:
It is a middleware so it don't require an abstraction